rs-matter/src/respond.rs

```diff
@@ -0,0 +1,269 @@
+pub struct CompositeExchangeHandler<F, S>(pub F, pub S);
+
+impl<F, S> ExchangeHandler for CompositeExchangeHandler<F, S>
+where
+    F: ExchangeHandler,
+    S: ExchangeHandler,
+{
+    async fn handle(&self, exchange: &mut Exchange<'_>) -> Result<(), Error> {
+        let result = self.0.handle(exchange).await;
+
+        match result {
+            Err(e) if e.code() == ErrorCode::InvalidProto => self.1.handle(exchange).await,
+            other => other,
+        }
+    }
+}
+
+/// A generic responder utility for accepting and handling exchanges received by the provided `Matter` stack,
+/// by applying the provided `ExchangeHandler` instance to each accepted exchange.
+///
+/// This responder uses an intra-task concurrency model - without an external executor - where all handling is done as a single future.
+pub struct Responder<'a, T> {
+    name: &'a str,
+    handler: T,
+    matter: &'a Matter<'a>,
+    respond_after_ms: u64,
+}
+
+impl<'a, T> Responder<'a, T>
+where
+    T: ExchangeHandler,
+{
+    /// Create a new responder.
+    ///
+    /// The `respond_after_ms` parameter instructs the responder how much time to wait before accepting an exchange.
+    ///
+    /// This is useful when utilizing multiple responders on a single `Matter` instance, where e.g. the first (main) responder is the actual one,
+    /// responsible for handling the incoming exchanges, while e.g. another one - with a non-zero `respond_after_ms` - is answerring all exchanges
+    /// not accepted in time by the main responder with a simple "I'm busy, try again later" handling.
+    #[inline(always)]
+    pub const fn new(
+        name: &'a str,
+        handler: T,
+        matter: &'a Matter<'a>,
+        respond_after_ms: u64,
+    ) -> Self {
+        Self {
+            name,
+            handler,
+            matter,
+            respond_after_ms,
+        }
+    }
+
+    /// Get a reference to the `ExchangeHandler` instance used by this responder
+    pub fn handler(&self) -> &T {
+        &self.handler
+    }
+
+    /// Run the responder with a given number of handlers.
+    pub async fn run<const N: usize>(&self) -> Result<(), Error> {
+        info!("{}: Creating {N} handlers", self.name);
+
+        let mut handlers = heapless::Vec::<_, N>::new();
+        info!(
+            "{}: Handlers size: {}B",
+            self.name,
+            core::mem::size_of_val(&handlers)
+        );
+
+        for index in 0..N {
+            let handler_id = (index as u8) + 2;
+
+            handlers
+                .push(self.handle(handler_id))
+                .map_err(|_| ())
+                .unwrap();
+        }
+
+        select_slice(&mut handlers).await.0
+    }
+
+    #[inline(always)]
+    async fn handle(&self, handler_id: impl Display) -> Result<(), Error> {
+        loop {
+            let _ = self.respond_once(&handler_id).await;
```

Comment:
this never propagates the error. should this be await? or is this on purpose like this? I read it as an infinte loop that intentionally ignores the result data.

Reply:
The logging is done in `respond_once`. Not sure we should move it here. I've documented why the error is not handled thought. I've kept the original signature returning `Result` though as we might decide to change our minds as to what we errors we propagate and what we log later.